src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.State.cs

```diff
@@ -90,6 +106,25 @@ protected State(ValueSource<Optional<Compilation>>? compilation, Compilation? de
                         ? new WeakValueSource<Compilation>(compilation)
                         : (ValueSource<Optional<Compilation>>)new ConstantValueSource<Optional<Compilation>>(compilation);
                 }
+
+                public static ConditionalWeakTable<ISymbol, ISymbol> GetCompilationAssembliesAndModules(Compilation compilation)
+                {
+                    var result = new ConditionalWeakTable<ISymbol, ISymbol>();
+
+                    var compAssembly = compilation.Assembly;
+                    result.Add(compAssembly, compAssembly);
+
+                    foreach (var reference in compilation.References)
+                    {
+                        var symbol = compilation.GetAssemblyOrModuleSymbol(reference);
+                        if (symbol == null)
+                            continue;
+
+                        result.Add(symbol, symbol);
+                    }
```

Comment:
@tmat it's the addition of this loop and the call to `compilation.GetAssemblyOrModuleSymbol` that breaks things.  I really can't tell why.  I've spent hours deep in the reference manager, but it's all incredibly cryptic to me.  One thing i *have* seen is that with this change, we now go through a new path in the reference manger here:

```
                // We are reading the Reference Manager state outside of a lock by accessing 
                // IsBound and HasCircularReference properties.
                // Once isBound flag is flipped the state of the manager is available and doesn't change.
                // 
                // If two threads are building SourceAssemblySymbol and the first just updated 
                // set isBound flag to 1 but not yet set lazySourceAssemblySymbol,
                // the second thread may end up reusing the Reference Manager data the first thread calculated. 
                // That's ok since 
                // 1) the second thread would produce the same data,
                // 2) all results calculated by the second thread will be thrown away since the first thread 
                //    already acquired SymbolCacheAndReferenceManagerStateGuard that is needed to publish the data.

                // The given compilation is the first compilation that shares this manager and its symbols are requested.
                // Perform full reference resolution and binding.
                if (!IsBound && CreateAndSetSourceAssemblyFullBind(compilation))
                {
                    // we have successfully bound the references for the compilation
                }
                else if (!HasCircularReference)
                {
                    // Another compilation that shares the manager with the given compilation
                    // already bound its references and produced tables that we can use to construct 
                    // source assembly symbol faster. Unless we encountered a circular reference.
                    CreateAndSetSourceAssemblyReuseData(compilation);
                }
```

Specifically we go into the second branch because `IsBound` is true.  This doesn't happen without this change here.  This makes me think the second branch is faulty somehow and is not properly initializing the assembly symbol with the correct data that the first branch would have.

Reply:
You said (in a separate forum) it's breaking stuff in Script scenarios? Does this also need to be do something about Compilation.PreviousSubmission? Can symbols from the previous submission get wrapped and retargeted like anything else?

Reply:
let me rephrase: it's not 'breaking' in that some tests fail.  it's breaking in that the addition of this causes crashes in GetDiagnostics core in the compiler.

It may be hte case that i'm not properly keeping track of all symbols.  however, regardless of that, it shoudl not be hte case that calling this causes interactive scenarios to crash in the compiler.